File: apps/api/app/schemas/quality.py
"""Pydantic v2 schemas for the OHLCV data-quality report (FRA-9)."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class AnomalyPoint(BaseModel):
    """One detected anomaly on a single bar."""

    model_config = ConfigDict(from_attributes=True)

    time: datetime
    rule: str  # non_positive_price | high_lt_low | negative_volume | zero_volume | large_return
    detail: str | None = None


class QualityReport(BaseModel):
    """On-demand quality report for one (asset, source, [start, end]) window."""

    model_config = ConfigDict(from_attributes=True)

    asset_id: uuid.UUID
    source: str
    start: date
    end: date
    expected_sessions: int
    observed_sessions: int
    missing_sessions_count: int
    coverage: float  # observed/expected in [0, 1]; 0.0 when expected == 0
    missing_sessions: list[date]
    anomalies: list[AnomalyPoint]

File: apps/web/src/api/quality.ts
/**
 * Data-quality client (FRA-9 endpoint).
 *
 * `GET /quality/{asset_id}?source=&start=&end=` → a QualityReport describing
 * coverage, missing sessions, and rule-based anomalies. 404 when the asset is
 * unknown; 422 when start>end or no trading calendar exists for the exchange.
 */
import { apiClient } from './client';
import type { QualityReport } from '@/types/api';

export interface FetchQualityParams {
  asset_id: string;
  source: string;
  start: string;
  end: string;
}

/** `GET /quality/{asset_id}` → quality report for the window. */
export async function fetchQuality(params: FetchQualityParams): Promise<QualityReport> {
  const { data } = await apiClient.get<QualityReport>(
    `/quality/${encodeURIComponent(params.asset_id)}`,
    {
      params: {
        source: params.source,
        start: params.start,
        end: params.end,
      },
    },
  );
  return data;
}
